File: apps/api/app/models/apiary.py
"""Apiary model."""

import uuid
from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class Apiary(Base):
    __tablename__ = "apiaries"

    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    latitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    longitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    city: Mapped[str | None] = mapped_column(String(255), nullable=True)
    country_code: Mapped[str | None] = mapped_column(String(10), nullable=True)
    hex_color: Mapped[str | None] = mapped_column(String(7), nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    archived_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    # Relationships
    user: Mapped["User"] = relationship("User", back_populates="apiaries")
    hives: Mapped[list["Hive"]] = relationship(
        "Hive", back_populates="apiary", cascade="all, delete-orphan"
    )
    tasks: Mapped[list["Task"]] = relationship("Task", back_populates="apiary")

File: apps/mobile/database/models/Apiary.ts
import { Model } from "@nozbe/watermelondb";
import { text, field, date, children, readonly } from "@nozbe/watermelondb/decorators";
import type { Associations } from "@nozbe/watermelondb/Model";

export default class Apiary extends Model {
  static table = "apiaries";

  static associations: Associations = {
    hives: { type: "has_many" as const, foreignKey: "apiary_id" },
    tasks: { type: "has_many" as const, foreignKey: "apiary_id" },
  };

  @text("name") name!: string;
  @field("latitude") latitude!: number | null;
  @field("longitude") longitude!: number | null;
  @text("city") city!: string | null;
  @text("country_code") countryCode!: string | null;
  @text("hex_color") hexColor!: string | null;
  @text("notes") notes!: string | null;
  @field("archived_at") archivedAt!: number | null;
  @readonly @date("created_at") createdAt!: Date;
  @readonly @date("updated_at") updatedAt!: Date;

  @children("hives") hives!: any;
  @children("tasks") tasks!: any;
}
